src/libaktualizr/primary/aktualizr_test.cc

```diff
@@ -492,6 +492,355 @@ TEST(Aktualizr, FullWithUpdatesNeedReboot) {
+
+    std::vector<std::pair<Uptane::EcuSerial, data::InstallationResult>> ecu_installation_res;
+    EXPECT_TRUE(storage->loadEcuInstallationResults(&ecu_installation_res));
+    EXPECT_EQ(ecu_installation_res.size(), 2);
+
+    for (const auto& ecu_install_res : ecu_installation_res) {
+      auto ecu_id = ecu_install_res.first.ToString();
+
+      if (ecu_id == primary_ecu_id) {
+        EXPECT_EQ(ecu_install_res.second.result_code.num_code, data::ResultCode::Numeric::kNeedCompletion);
+        EXPECT_EQ(ecu_install_res.second.success, false);
+        EXPECT_EQ(ecu_install_res.second.description, "Application successful, need reboot");
+
+      } else if (ecu_id == secondary_ecu_id) {
+        EXPECT_EQ(ecu_install_res.second.result_code.num_code, data::ResultCode::Numeric::kOk);
+        EXPECT_EQ(ecu_install_res.second.success, true);
+        EXPECT_EQ(ecu_install_res.second.description, "");
+
+      } else {
+        FAIL() << "Unexpected ECU serial: " << ecu_install_res.first;
+      }
+    }
+
+    ASSERT_TRUE(
+        storage->loadInstalledVersions(primary_ecu_id, &installed_versions, &current_version, &pending_version));
+    ASSERT_EQ(installed_versions.size(), 1);
+    // why valid
+    ASSERT_TRUE(installed_versions[0].IsValid());
+    // why 0 if should be 1/true
+    ASSERT_EQ(pending_version, 0);
+    ASSERT_EQ(current_version, 0);
+
+    ASSERT_TRUE(
+        storage->loadInstalledVersions(secondary_ecu_id, &installed_versions, &current_version, &pending_version));
+    ASSERT_EQ(installed_versions.size(), 1);
+    ASSERT_TRUE(installed_versions[0].IsValid());
+    ASSERT_EQ(pending_version, 0);
+    // why 0, this is secondary that was succesfully installed, so it should 1/true ???
+    ASSERT_EQ(current_version, 0);
+  }
+
+  {
+    // now try to finalize the previously installed updates with enabled failure injection
+    fiu_init(0);
+    fiu_enable("fake_install_finalization_failure", 1, nullptr, 0);
+    http_server_mock->clearReportEvents();
+
+    Aktualizr aktualizr(conf, storage, http_server_mock);
+    EventHandler event_hdlr(aktualizr);
+    aktualizr.Initialize();
+    aktualizr.UptaneCycle();
+
+    fiu_disable("fake_install_finalization_failure");
+
+    result::UpdateCheck update_res = aktualizr.CheckUpdates().get();
+    // Is it correct expectations ?????
+    // it fails because it sees that aktualizr got higher version (2) than the one that server returns 1
+    // moreover it takes version out of the stored metadata but not from successfully installed target
+    // see sotauptaneclient.cc::413
+    ASSERT_EQ(update_res.status, result::UpdateStatus::kError);
+    ASSERT_FALSE(aktualizr.uptane_client_->hasPendingUpdates());
+    ASSERT_TRUE(event_hdlr.checkRecievedEvents({"UpdateCheckComplete"}));
+    ASSERT_TRUE(http_server_mock->checkRecievedReports({"EcuInstallationCompleted"}));
+    ASSERT_FALSE(http_server_mock->wasInstallSuccessfull());
```

Comment:
To be fair, this is kind of a weird case. So normally, finalization failure would imply that we couldn't boot the newly installed version, so we either rolled back to the previous version or are stuck in an unknown state. The fact that we faked failure here means that we are in a state that probably shouldn't be possible. If you are going to fake failure, then we should probably force libaktualizr to "roll back" to the previous version. Then perhaps this would make more sense.

Reply:
If a failure at the installation finalization has happened then a bootloader had switched back to the previous ostree-based revision of rootfs at boot time before running aktualizr, so we don't need to emulate "rolling back" here.

Reply:
But then the version shouldn't be 2, if I understand correctly. We should have rolled back to running version 1.

Reply:
Yes version should not be 2, but still aktualizr doesn't do any roll back related actions also it actually does know what was the previous version since installation results' life span just within UptaneCycle. 

In this specific case the 'version' is obtained from director's targets metadata what was successfully stored and has never been removed from DB in spite the fact that their installation has failed :).

CheckUpdates()->fetchMeta()->updateMeta()->uptaneIteration()->updateDirectorMeta()->{Update Director Targets Metadata}

Reply:
Removed CheckUpdates() since it's out of scope of this test.